File: android/app/src/main/java/com/dawainnovation/lens/NativeBiometricPlugin.kt
package com.dawainnovation.lens

import androidx.biometric.BiometricManager
import androidx.biometric.BiometricManager.Authenticators.BIOMETRIC_STRONG
import androidx.biometric.BiometricManager.Authenticators.DEVICE_CREDENTIAL
import androidx.biometric.BiometricPrompt
import androidx.core.content.ContextCompat
import com.getcapacitor.*
import com.getcapacitor.annotation.CapacitorPlugin

@CapacitorPlugin(name = "NativeBiometric")
class NativeBiometricPlugin : Plugin() {

    @PluginMethod
    fun isAvailable(call: PluginCall) {
        val manager = BiometricManager.from(context)
        val result = JSObject()
        when (manager.canAuthenticate(BIOMETRIC_STRONG or DEVICE_CREDENTIAL)) {
            BiometricManager.BIOMETRIC_SUCCESS -> {
                result.put("available", true)
                // Distinguish biometric hardware type
                val type = when {
                    manager.canAuthenticate(BIOMETRIC_STRONG) == BiometricManager.BIOMETRIC_SUCCESS -> "fingerprint"
                    else -> "device_credential"
                }
                result.put("biometryType", type)
                call.resolve(result)
            }
            else -> {
                result.put("available", false)
                result.put("biometryType", "none")
                call.resolve(result)
            }
        }
    }

    @PluginMethod
    fun authenticate(call: PluginCall) {
        val reason = call.getString("reason") ?: "Verify your identity"
        @Suppress("UNUSED_VARIABLE")
        val fallbackTitle = call.getString("fallbackTitle") ?: "Use Passcode"

        // BiometricPrompt MUST be constructed and shown on the UI thread
        // from a FragmentActivity — Capacitor's `activity` is AppCompatActivity (a FragmentActivity).
        activity.runOnUiThread {
            val executor = ContextCompat.getMainExecutor(context)

            val callback = object : BiometricPrompt.AuthenticationCallback() {
                override fun onAuthenticationSucceeded(result: BiometricPrompt.AuthenticationResult) {
                    val res = JSObject()
                    res.put("success", true)
                    call.resolve(res)
                }

                override fun onAuthenticationFailed() {
                    // User attempted but biometric did not match — let them retry;
                    // do NOT resolve yet.
                }

                override fun onAuthenticationError(errorCode: Int, errString: CharSequence) {
                    // Resolve (not reject) so JS can handle errors gracefully
                    // (e.g. user cancelled, lockout, no hardware).
                    val res = JSObject()
                    res.put("success", false)
                    res.put("error", errString.toString())
                    call.resolve(res)
                }
            }

            val prompt = BiometricPrompt(activity, executor, callback)
            val info = BiometricPrompt.PromptInfo.Builder()
                .setTitle("Dawa Lens")
                .setSubtitle(reason)
                // DEVICE_CREDENTIAL allows PIN/pattern/password as fallback;
                // setNegativeButtonText must NOT be set when using DEVICE_CREDENTIAL.
                .setAllowedAuthenticators(BIOMETRIC_STRONG or DEVICE_CREDENTIAL)
                .build()

            prompt.authenticate(info)
        }
    }
}

File: ios/App/App/NativeBiometricPlugin.swift
import Foundation
import Capacitor
import LocalAuthentication

@objc(NativeBiometricPlugin)
public class NativeBiometricPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "NativeBiometricPlugin"
    public let jsName = "NativeBiometric"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "isAvailable", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "authenticate", returnType: CAPPluginReturnPromise),
    ]

    @objc func isAvailable(_ call: CAPPluginCall) {
        let context = LAContext()
        var error: NSError?
        let canEvaluate = context.canEvaluatePolicy(.deviceOwnerAuthenticationWithBiometrics, error: &error)
        var result = JSObject()
        result["available"] = canEvaluate
        if canEvaluate {
            switch context.biometryType {
            case .faceID:  result["biometryType"] = "face"
            case .touchID: result["biometryType"] = "fingerprint"
            case .opticID: result["biometryType"] = "opticid"
            default:       result["biometryType"] = "none"
            }
        } else {
            result["biometryType"] = "none"
        }
        call.resolve(result)
    }

    @objc func authenticate(_ call: CAPPluginCall) {
        let reason = call.getString("reason") ?? "Verify your identity to access your health data"
        let context = LAContext()
        context.localizedFallbackTitle = call.getString("fallbackTitle") ?? "Use Passcode"

        // .deviceOwnerAuthentication falls back to passcode when biometrics fail or are unavailable
        context.evaluatePolicy(.deviceOwnerAuthentication, localizedReason: reason) { success, error in
            var result = JSObject()
            result["success"] = success
            if let err = error {
                result["error"] = err.localizedDescription
            }
            call.resolve(result)
        }
    }
}
